model/labels/labels_internlabels.go

```diff
@@ -0,0 +1,778 @@
+			break
+		}
+		if name == names[j] {
+			b = append(b, name...)
+			b = append(b, seps[0])
+			b = append(b, value...)
+			b = append(b, seps[0])
+		}
+	}
+
+	return xxhash.Sum64(b), b
+}
+
+// HashWithoutLabels returns a hash value for all labels except those matching
+// the provided names.
+// 'names' have to be sorted in ascending order.
+func (ls Labels) HashWithoutLabels(b []byte, names ...string) (uint64, []byte) {
+	b = b[:0]
+	j := 0
+	for i := 0; i < len(ls.data); {
+		var name, value string
+		name, i = decodeString(ls.syms, ls.data, i)
+		value, i = decodeString(ls.syms, ls.data, i)
+		for j < len(names) && names[j] < name {
+			j++
+		}
+		if name == MetricName || (j < len(names) && name == names[j]) {
+			continue
+		}
+		b = append(b, name...)
+		b = append(b, seps[0])
+		b = append(b, value...)
+		b = append(b, seps[0])
+	}
+	return xxhash.Sum64(b), b
+}
+
+// BytesWithLabels is just as Bytes(), but only for labels matching names.
+// 'names' have to be sorted in ascending order.
+func (ls Labels) BytesWithLabels(buf []byte, names ...string) []byte {
+	b := bytes.NewBuffer(buf[:0])
+	j := 0
+	for pos := 0; pos < len(ls.data); {
+		lName, newPos := decodeString(ls.syms, ls.data, pos)
+		lValue, newPos := decodeString(ls.syms, ls.data, newPos)
+		for j < len(names) && names[j] < lName {
+			j++
+		}
+		if j == len(names) {
+			break
+		}
+		if lName == names[j] {
+			if b.Len() > 1 {
+				b.WriteByte(seps[0])
+			}
+			b.WriteString(lName)
+			b.WriteByte(seps[0])
+			b.WriteString(lValue)
+		}
+		pos = newPos
+	}
+	return b.Bytes()
+}
+
+// BytesWithoutLabels is just as Bytes(), but only for labels not matching names.
+// 'names' have to be sorted in ascending order.
+func (ls Labels) BytesWithoutLabels(buf []byte, names ...string) []byte {
+	b := bytes.NewBuffer(buf[:0])
+	j := 0
+	for pos := 0; pos < len(ls.data); {
+		lName, newPos := decodeString(ls.syms, ls.data, pos)
+		lValue, newPos := decodeString(ls.syms, ls.data, newPos)
+		for j < len(names) && names[j] < lName {
+			j++
+		}
+		if j == len(names) || lName != names[j] {
+			if b.Len() > 1 {
+				b.WriteByte(seps[0])
+			}
+			b.WriteString(lName)
+			b.WriteByte(seps[0])
+			b.WriteString(lValue)
+		}
+		pos = newPos
+	}
+	return b.Bytes()
+}
+
+// Copy returns a copy of the labels.
+func (ls Labels) Copy() Labels {
+	buf := append([]byte{}, ls.data...)
+	return Labels{syms: ls.syms, data: yoloString(buf)}
+}
+
+// Get returns the value for the label with the given name.
+// Returns an empty string if the label doesn't exist.
+func (ls Labels) Get(name string) string {
+	if name == "" { // Avoid crash in loop if someone asks for "".
+		return "" // Prometheus does not store blank label names.
+	}
+	for i := 0; i < len(ls.data); {
+		var lName, lValue string
+		lName, i = decodeString(ls.syms, ls.data, i)
+		if lName == name {
+			lValue, _ = decodeString(ls.syms, ls.data, i)
+			return lValue
+		} else if lName[0] > name[0] { // Stop looking if we've gone past.
```

Comment:
Are we checking only `[0]` as a performance optimization? Does this play well with Unicode (seems like U nicode support is coming?)

Is it really faster than `lName > name`? (I guess CPUs have that optimized)

Reply:
This one is significantly faster on the benchmark:

With `[0]`:
```
BenchmarkLabels_Get/with_5_labels/not-found_label/get-4         50844589                23.60 ns/op            0 B/op          0 allocs/op
BenchmarkLabels_Get/with_5_labels/not-found_label/has-4         49705240                23.91 ns/op            0 B/op          0 allocs/op
BenchmarkLabels_Get/with_10_labels/not-found_label/get-4        49250290                23.99 ns/op            0 B/op          0 allocs/op
BenchmarkLabels_Get/with_10_labels/not-found_label/has-4        50232136                23.74 ns/op            0 B/op          0 allocs/op
BenchmarkLabels_Get/with_30_labels/not-found_label/get-4        51744460                24.15 ns/op            0 B/op          0 allocs/op
BenchmarkLabels_Get/with_30_labels/not-found_label/has-4        49727462                24.52 ns/op            0 B/op          0 allocs/op
```

Without:
```
BenchmarkLabels_Get/with_5_labels/not-found_label/get-4         29063010                39.48 ns/op            0 B/op          0 allocs/op
BenchmarkLabels_Get/with_5_labels/not-found_label/has-4         30483260                45.85 ns/op            0 B/op          0 allocs/op
BenchmarkLabels_Get/with_10_labels/not-found_label/get-4        27639243                46.17 ns/op            0 B/op          0 allocs/op
BenchmarkLabels_Get/with_10_labels/not-found_label/has-4        30595272                46.34 ns/op            0 B/op          0 allocs/op
BenchmarkLabels_Get/with_30_labels/not-found_label/get-4        30075024                40.38 ns/op            0 B/op          0 allocs/op
BenchmarkLabels_Get/with_30_labels/not-found_label/has-4        29790415                39.70 ns/op            0 B/op          0 allocs/op
```